### PR TITLE
[None][refactor] Organize SMG gRPC adapter by protocol

### PR DESCRIPTION
**Pull request by QiJune**
## Dev Engineer Review

- Refactored the SMG gRPC adapter into `tensorrt_llm.grpc.smg`.
- Added `bindings.py`, `server.py`, and package metadata.
- Updated protobuf imports to use the new `bindings` path.
- Updated `serve.py` to launch `launch_smg_server`.
- Added optional dependency handling and actionable installation errors.
- Added server cleanup for binding and startup failures.
- Added `protobuf>=5.27.2` to runtime requirements.
- Added `smg-grpc-proto>=0.4.2` to development requirements.
- Compatibility aliases for removed gRPC exports and module paths remain a follow-up.
- The CLI help and error text use `tensorrt_llm[devel]`. The required package name is `tensorrt_llm[grpc-smg]`. Align the documentation and error text.
- Document that `trtllm-serve --grpc` requires `pip install tensorrt_llm[grpc-smg]`.
- Confirm that `ImportError` handling only catches missing `smg_grpc_proto` and does not mask unrelated import failures.
- Confirm that binding and startup failure paths always stop the gRPC server and shut down the loaded engine.
- Confirm that `tests/unittest/grpc/smg/test_smg.py` is registered in the required CI test list.
- Review the reported CI failures, including the test-list failure involving synthetic cache-transceiver entries and failures after the optional packaging change.

## QA Engineer Review

- Added:
  - `test_smg_bindings_missing_gives_actionable_error`
  - `test_smg_bindings_preserves_unrelated_import_error`
  - `test_smg_bindings_present_smoke`
  - `test_smg_server_startup_failure_cleans_up`
- Updated `tests/unittest/grpc/smg/test_smg.py` for the new package path and optional dependency behavior.
- Added `unittest/grpc/test_grpc_optional.py` and `unittest/grpc/smg/test_smg.py` to `tests/integration/test_lists/test-db/l0_a10.yml`.
- Test coverage is registered for the A10 PyTorch pre-merge list.
- Verdict: needs follow-up until CI execution and the optional package-name documentation are confirmed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by BowenFu** [comment]
The server move appears behavior-preserving: message limits, keepalive settings, signal handling, and shutdown ordering are unchanged. Could you address three follow-ups before merge?

- Register `tests/unittest/grpc/smg/test_smg.py` in CI.
- Consider one-release aliases for the old gRPC exports/module paths if router-side code may use them.
- Extend cleanup around binding/startup so failures after loading the LLM still stop the server and engine.

I’d wait for a clean pipeline before approval.

**Comment by BowenFu** [comment]
The current red runs look unrelated: build 51350 reports 27,211 passing tests and only stage-level failures. Several PRs on the same base also fail Check Test List on synthetic cache-transceiver precheck names in `waives.txt`.

Could you rerun once the base validator issue is fixed? My remaining code concerns are CI registration for `test_smg.py` and compatibility aliases for the removed gRPC paths.

**Comment by BowenFu** [comment]
`smg-grpc-proto` becoming optional looks reasonable. Could you tighten two edges before merge?

- Document that `trtllm-serve --grpc` now requires `pip install tensorrt_llm[grpc-smg]`.
- Narrow both `ImportError` handlers to the missing top-level `smg_grpc_proto` package. Currently, unrelated gRPC/protobuf or sibling-import failures are rewritten as an installation hint, which could hide real breakage. Matching `e.name` or checking `find_spec` would preserve the helpful message without masking other errors.

The title could also mention optional packaging, since this is more than a directory refactor. I’d hold until the import boundary is narrowed and CI completes.

File: setup.py [hunk 1]
```diff
@@ -130,6 +130,10 @@ def has_ext_modules(self):
     Path("requirements-dev-windows.txt"
          if on_windows else "requirements-dev.txt"))
 mx_deps = ["modelexpress==0.4.1"]
+# Optional SMG gRPC adapter dependency. Not part of the default install; opt in
+# with `pip install tensorrt_llm[grpc-smg]`. Also kept in requirements-dev.txt so
+# the adapter is present (and tested) in every CI/dev environment.
+grpc_smg_deps = ["smg-grpc-proto>=0.4.2"]
```
Comment (tburt-nv) [inline setup.py]: This is already part of the `devel` group; can we we remove this `grpc-smg` group and rely on the `devel` group instead? If we add many fine-grained optional groups, it will become difficult to test important combinations of these groups.

```suggestion
```

File: tensorrt_llm/grpc/smg/bindings.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+"""Generated protobuf bindings used by the SMG adapter."""
+
+try:
+    from smg_grpc_proto.generated import trtllm_service_pb2, trtllm_service_pb2_grpc
+except ImportError as e:
+    raise ImportError(
+        "The SMG gRPC adapter requires the optional 'smg-grpc-proto' package, "
+        "which is not part of the default TensorRT-LLM installation. Install it "
+        "with: pip install tensorrt_llm[grpc-smg]"
+    ) from e
```
Comment (tburt-nv) [inline tensorrt_llm/grpc/smg/bindings.py:28]: Are there any workflows in the docs that would exercise this path? If so, can we update them to install `tensorrt_llm[devel]` ahead of time?


```suggestion
try:
    from smg_grpc_proto.generated import trtllm_service_pb2, trtllm_service_pb2_grpc
except ImportError as e:
    raise ImportError(
        "The SMG gRPC adapter requires the optional 'smg-grpc-proto' package, "
        "which is not part of the default TensorRT-LLM installation. Install it "
        "with: pip install tensorrt_llm[devel]"
    ) from e
```

File: tensorrt_llm/grpc/smg/server.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            if reflection is not None:
+                service_names = (
+                    trtllm_service_pb2.DESCRIPTOR.services_by_name["TrtllmService"].full_name,
+                    reflection.SERVICE_NAME,
+                )
+                reflection.enable_server_reflection(service_names, server)
+                logger.info("gRPC reflection enabled")
+
+            address = f"{host}:{port}"
+            server.add_insecure_port(address)
```
Comment (brnguyen2) [inline tensorrt_llm/grpc/smg/server.py:106]: Non-blocking, and pre-existing rather than introduced here — fine as a follow-up PR.

`add_insecure_port` never raises on failure — it returns the bound port, or **0** when binding fails (port already in use, bad address). The return value is ignored here, so on a real bind failure the server "starts" cleanly, logs `Server is ready to accept requests`, and blocks on `stop_event` while listening on nothing. Since this PR already adds the failure-cleanup path, checking the return is a natural next step:

```python
if server.add_insecure_port(address) == 0:
    raise RuntimeError(f"Failed to bind gRPC server to {address}")
```

The `finally` block then handles teardown correctly.

File: tests/unittest/grpc/test_grpc_optional.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    )
+    from tensorrt_llm.grpc.smg import server as server_module
+
+    llm = MagicMock()
+    grpc_server = MagicMock()
+    grpc_server.start = AsyncMock()
+    grpc_server.stop = AsyncMock()
+
+    if failure_point == "bind":
+        grpc_server.add_insecure_port.side_effect = RuntimeError("bind failed")
```
Comment (brnguyen2) [inline tests/unittest/grpc/test_grpc_optional.py:108]: Follow-up along with the `server.py` bind check — no need to block this PR.

This simulates a bind failure with a raising mock, but real `grpc.aio` `add_insecure_port` doesn't raise — it returns 0 on failure. As written, the "bind" case validates cleanup for a failure mode that can't occur, while the actual bind-failure behavior (silent unbound server) goes untested. Once `server.py` checks the return value, change this case to `grpc_server.add_insecure_port.return_value = 0` and assert on the resulting error, so the test exercises the real contract.

File: tensorrt_llm/grpc/smg/request_manager.py [hunk 5]
```diff
@@ -211,8 +211,18 @@ def get_model_config(self) -> Dict[str, Any]:
...
             # Try to get tokenizer info
             if hasattr(self.llm, "tokenizer") and self.llm.tokenizer is not None:
-                if hasattr(self.llm.tokenizer, "vocab_size"):
-                    config["vocab_size"] = self.llm.tokenizer.vocab_size
+                # TransformersTokenizer wraps the HF tokenizer without
+                # overriding vocab_size, so the wrapper attribute resolves to
+                # PreTrainedTokenizerBase's abstract property (raises
+                # NotImplementedError). Read from the inner tokenizer.
+                tokenizer = self.llm.tokenizer
+                inner_tokenizer = getattr(tokenizer, "tokenizer", tokenizer)
```
Comment (brnguyen2) [inline tensorrt_llm/grpc/smg/request_manager.py:219]: This is a behavior fix (GetModelInfo previously hit `PreTrainedTokenizerBase`'s `NotImplementedError` through the wrapper and reported no vocab_size), not part of the package reorganization — worth a line in the PR description so it's findable later. Also, none of the tests in `test_smg.py` cover `get_model_config()` with a `TransformersTokenizer`-wrapped LLM; a small unit test asserting `vocab_size` comes from the inner tokenizer would keep this from regressing.

File: setup.py (hunk 1 above)
Comment (QiJune) [reply setup.py]: The `grpc-smg` extra is deliberate rather than incidental granularity — one extra per gateway adapter, which is the ownership/packaging principle we aligned on with the Dynamo team for the OpenEngine contribution (#17084). Three reasons `devel` alone doesn't work:

1. **It's user-facing install surface, not a dev knob.** `trtllm-serve --grpc` is a production entry point; asking a deployment image to `pip install tensorrt_llm[devel]` pulls the entire dev toolchain just to get one proto package. `grpc-smg` installs exactly `smg-grpc-proto`.
2. **The second gateway can't live in `devel`.** OpenEngine's bindings pin protobuf 6.x (and cap grpcio<2) while SMG pins 5.x — they can conflict. If both sat in `devel`, `pip install tensorrt_llm[devel]` would break for everyone the day #17084 lands. Per-gateway extras scope the conflict to users who explicitly request both.
3. **The combination-explosion concern is bounded by construction.** CI installs exactly zero or one `requirements-<gateway>.txt` per test stage (see the comment in `L0_Test.groovy`), so tested combinations grow linearly with gateways, not combinatorially.

Note `devel` still includes the SMG pins in this PR, so existing dev workflows are unchanged; `grpc-smg` is the narrow production-facing alias. vLLM ships its SMG-based gRPC server the same way — a `"grpc"` extra among its peer extras.

File: tensorrt_llm/grpc/smg/bindings.py (hunk 2 above)
Comment (QiJune) [reply tensorrt_llm/grpc/smg/bindings.py:28]: The user-facing workflow is `trtllm-serve --grpc`, and this PR makes all three touchpoints consistent on the narrow extra: the CLI help text ("Requires the tensorrt_llm[grpc-smg] extra"), the `find_spec` preflight in `serve.py`, and this bindings error all say `pip install "tensorrt_llm[grpc-smg]"`. We kept the hint on `grpc-smg` rather than `devel` intentionally: it's the minimal, production-safe instruction, and since `devel` is a superset of `grpc-smg`, anyone who already installed `[devel]` never sees this error. The release container also now installs `[mx,grpc-smg]` (Dockerfile.multi), so containerized users are covered by default. I'll add a note next to the `--grpc` docs mentioning the extra.

**Comment by QiJune** [comment]
@CatherineSue Heads-up on a packaging change to the SMG gRPC adapter, since you've been the main contributor on the SMG side.

**What changes.** This PR moves `smg-grpc-proto` out of TensorRT-LLM's default install: it is no longer in `requirements.txt`, and the adapter code now lives under `tensorrt_llm/grpc/smg/`. Installation becomes opt-in:

```bash
pip install "tensorrt_llm[grpc-smg]"
```

This applies to the NGC release container as well — the image no longer preinstalls the SMG dependency, so existing containerized `trtllm-serve --grpc` deployments need to add the one-line `pip install` above on top of the image. Running `--grpc` without the extra fails fast with an error message pointing at that exact command.

**Why.** A second community gateway adapter (OpenEngine, #17084, from the Dynamo team) is being contributed alongside SMG, and the two bindings pin protobuf incompatibly (5.x vs 6.x). The principle we've aligned on is that gateway adapters are peer, community-maintained, opt-in extras: no gateway sits in the default install or the release image, and each owns a `requirements-<gateway>.txt` as the single source of truth for its pins. SMG is simply the first adapter moved onto this model, and OpenEngine will be held to the same bar.

**What stays the same.** Runtime behavior of the adapter is unchanged — this PR is a relocation plus the packaging split. The SMG unit tests remain in pre-merge CI (`tests/unittest/grpc/smg/`), with the dependency provisioned explicitly in the test environments.

If SMG-side docs reference the old install assumption ("gRPC works out of the box with `pip install tensorrt_llm`") or the old module paths (`tensorrt_llm.grpc.grpc_servicer` etc.), they'll want an update — module-path compatibility aliases are tracked as a follow-up.

**Comment by QiJune** [comment]
/bot skip --comment "All smg test passed, the failures are unrelated."
